### PR TITLE
[Security] Improve deprecation logic to minimalize the changes required in 3.0

### PR DESCRIPTION
**Pull request by wouterj**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

File: src/Symfony/Component/Security/Core/Authorization/Voter/AbstractVoter.php [hunk 1]
```diff
@@ -146,6 +139,6 @@ protected function isGranted($attribute, $object, $user = null)
      */
     protected function voteOnAttribute($attribute, $object, TokenInterface $token)
     {
-        return false;
+        return $this->isGranted($attribute, $object, $token->getUser());
```
Comment (Tobion) [inline src/Symfony/Component/Security/Core/Authorization/Voter/AbstractVoter.php]: Can we not move the `trigger_error` here as well? Then we don't need reflection at all.

Comment (Tobion) [inline src/Symfony/Component/Security/Core/Authorization/Voter/AbstractVoter.php]: Hm classes that don't support any attributes would not trigger a warning then.

**Comment by Tobion** [comment]
:+1:

**Comment by nicolas-grekas** [comment]
:+1:

**Comment by stof** [comment]
you should also add a deprecation warning in ``AbstractVoter::isGranted, in case someone overwrites ``voteOnAttribute`` but still calls ``isGranted``

**Comment by WouterJ** [comment]
@stof I don't like that, as bundles supporting 2.7 and 2.8 will probably create both methods and call `isGranted` from `voteOnAttribute`, like the BC layer implemented in this PR.

**Pull request closed without merging**